lib/datadog/di/probe_manager.rb

```diff
@@ -0,0 +1,182 @@
+
+        raise
+      end
+
+      def remove_other_probes(probe_ids)
+        pending_probes.values.each do |probe|
+          unless probe_ids.include?(probe.id)
+            pending_probes.delete(probe.id)
+          end
+        end
+        installed_probes.values.each do |probe|
+          unless probe_ids.include?(probe.id)
+            begin
+              instrumenter.unhook(probe)
+              # Only remove the probe from installed list if it was
+              # successfully de-instrumented. Active probes do incur overhead
+              # for the running application, and if the error is ephemeral
+              # we want to try removing the probe again at the next opportunity.
+              #
+              # TODO give up after some time?
+              installed_probes.delete(probe.id)
+            rescue => exc
+              raise if settings.dynamic_instrumentation.internal.propagate_all_exceptions
+              # Silence all exceptions?
+              # TODO should we propagate here and rescue upstream?
+              logger.warn("Error removing probe #{probe.id}: #{exc.class}: #{exc}")
+              telemetry&.report(exc, description: "Error removing probe #{probe.id}")
+            end
+          end
+        end
+      end
+
+      private def install_pending_method_probes(cls)
+        # TODO search more efficiently than linearly
+        pending_probes.each do |probe_id, probe|
+          if probe.method?
+            # TODO move this stringification elsewhere
+            if probe.type_name == cls.name
+              # TODO is it OK to hook from trace point handler?
+              # TODO the class is now defined, but can hooking still fail?
+              # TODO pass rate_limiter here, need to get it from somewhere
+              hook_method(probe.type_name, probe.method_name,
+                rate_limiter: probe.rate_limiter, &instance_method(:probe_executed_callback))
+              pending_probes.delete(probe.id)
+              break
+            end
+          end
+        end
+      end
+
+      def install_pending_line_probes(file)
+        pending_probes.values.each do |probe|
+          if probe.line?
+            if probe.file_matches?(file)
+              add_probe(probe)
+            end
+          end
+        end
+      end
+
+      def probe_executed_callback(probe:, **opts)
+        unless probe.emitting_notified?
+          payload = probe_notification_builder.build_emitting(probe)
+          probe_notifier_worker.add_status(payload)
+          probe.emitting_notified = true
+        end
+
+        payload = probe_notification_builder.build_executed(probe, **opts)
+        probe_notifier_worker.add_snapshot(payload)
+      end
+
+      # Class/module definition trace point (:end type).
+      # Used to install hooks when the target classes/modules aren't yet
+      # defined when the hook request is received.
+      attr_reader :definition_trace_point
```

Comment:
Any reason this `attr_reader` is declared at the end of the class, instead of together with the other `attr_reader`s?

Reply:
It's an internal attribute, even though it is set from outside of the class.